sciope/inference/rep_smc_abc.py

```diff
@@ -0,0 +1,187 @@
+from sciope.inference.inference_base import InferenceBase
+from sciope.inference import abc_inference
+from sciope.core import core
+from sciope.utilities.distancefunctions import euclidean as euc
+from sciope.utilities.summarystats import burstiness as bs
+from sciope.utilities.housekeeping import sciope_logger as ml
+from sciope.utilities.priors.prior_base import PriorBase
+from sciope.utilities.epsilonselectors import RelativeEpsilonSelector
+
+import numpy as np
+import dask
+from dask.distributed import futures_of, as_completed, wait
+from dask import delayed
+
+class ReplenishmentSMCABC(InferenceBase):
+    """
+    Replenishment SMC - Approximate Bayesian Computation
+
+    Properties/variables:
+    * data                      (observed / fixed data)
+    * sim   					(simulator function handle)
+    * prior_function			(prior over the simulator parameters)
+    * perturbation_kernel       (kernel for perturbing parameters samples)
+    * summaries_function    	(summary statistics calculation function)
+    * distance_function         (function calculating deviation between simulated statistics and observed statistics)
+    * summaries_divisor         (numpy array of maxima - used for normalizing summary statistic values)
+    * use_logger    			(whether logging is enabled or disabled)
+
+    Methods:
+    * infer 					(perform parameter inference)
+    """
+
+    def __init__(self, data, sim, prior_function, perturbation_kernel,
+                 summaries_function=bs.Burstiness(),
+                 distance_function=euc.EuclideanDistance(),
+                 summaries_divisor=None, use_logger=False):
+
+        self.name = 'SMC-ABC'
+        super(ReplenishmentSMCABC, self).__init__(self.name, data, sim, use_logger)
+
+        self.prior_function = prior_function
+        self.summaries_function = summaries_function
+        self.distance_function = distance_function
+        self.summaries_divisor = summaries_divisor
+        self.perturbation_kernel = perturbation_kernel
+
+        if self.use_logger:
+            self.logger = ml.SciopeLogger().get_logger()
+            self.logger.info("Sequential Monte-Carlo Approximate Bayesian Computation initialized")
+
+    def _simulate_N(self, prior, N):
+
+        params_delayed = prior.draw(N)
+        params, = dask.compute(params_delayed, num_workers = 10)
+        params = np.vstack(params).reshape(N, -1)
+
+        results = []
+        for i in range(params.shape[0]):
+            current_param = params[i,:]
+            lazy_results = dask.delayed(self._simulate_distance)(current_param)
+            results.append(lazy_results)
+
+        computed_results, = dask.compute(results)
+
+        return np.asarray(computed_results), params
+
+    def _simulate_distance(self, param):
+        res = self.summaries_function(self.sim(param))
+        res_obs = self.summaries_function(self.data)
+        return self.distance_function.compute(res_obs, res)
```

Comment:
Replace by` sciope.core.get_graph_chunked()`, should support chunked data. We might need to change ` sciope.core.get_graph_chunked()` to handle a single param without a sampler function